### ISSUE TITLE
RPi4 - kodi 19.0 b1 - Dvd menu highlighting not working with some dvd's 

### ISSUE THREAD

**Issue by A1RF0RCE1** [xbmc/xbmc#18898]
### Describe the bug
Here is a clear and concise description of what the problem is:

When navigating a dvd menu no menu item highlight is visible, the selecting sound is hearable but no visible hightlight change is occurring. When pressing the remote, the selected menu item is carried out.  
Also when selecting a non working dvd when kodi just started no menu is visible at all (black screen), when pressing the remote in the blind the movie starts playing and the screen will appear. Selecting the movie for the second will show the dvd menu without highlighted menu items.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

Visible highlight change in the dvd menu

## Actual Behavior

No highlight visible at all

### To Reproduce

1. select a non working dvd 
2.
3.

### Debuglog

The debuglog can be found here:

https://paste.kodi.tv/taquwawuwi.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP
-  [X] Libreelec 9.8

 - Operating system version/name:
 - Kodi version:
19.0 beta 1

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by A1RF0RCE1** [comment]
Libreelec 9.2.6 with kodi 18.7 is working as expected.

**Comment by A1RF0RCE1** [comment]
It also occurs that a dvd with multiple menu's the first menu for example for the language selection doesn't have working menu highlights the second actual dvd menu shows proper highlights. It's like in some cases the menu isn't properly detected.

**Comment by A1RF0RCE1** [comment]
Disabling the option "Allow using DRM PRIME decoder" in the kodi player settings fixes the problem, so the bug is in there somewhere :-)

**Comment by A1RF0RCE1** [comment]
Changed: 
xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp

comment out 
// if (!(hints.codecOptions & CODEC_FORCE_SOFTWARE)) 

fixed the problem

**Comment by lrusak** [comment]
Just disable drmprime decoding. Or better yet use drmprime with EGL rendering.

**Comment by A1RF0RCE1** [comment]
Hi lukas, 

Thanks for the reply, just tested the drmprime with EGL rendering. This fixes the black screen on cold start but the dvd menu overlay is still broken. 
disabling drmprime gives a less smooter video playback on other movies (h264). 

I removed the if statement because the working dvd entered this section and the none working dvd wasn’t .

**Comment by A1RF0RCE1** [comment]
Thanks for the reply, just tested the drmprime with EGL rendering. This fixes the black screen on cold start but the menu overlay is still broken. 
disabling drmprime gives a less smooter video playback on other movies (h264)

**Comment by lrusak** [comment]
> disabling drmprime gives a less smooter video playback on other movies (h264).

This is because drm prime is your hardware decoding :smile:

**Comment by A1RF0RCE1** [comment]
I assumed it had something to do with that :-).

**Comment by dkadioglu** [comment]
I have the same problems with LibreELEC 10 nightly (self built) on RPi 4. Anything, I can provide to help to fix this?

**Comment by A1RF0RCE1** [comment]
Hi dkadioglu, i have created a small patch for kodi. giving the same behaviour as kodi 18 did for dvd’s seems to works for le >=10 as well

**Comment by dkadioglu** [comment]
Hi @A1RF0RCE1 thanks for the hint, now it works! However, are there any side effects when playing any other non-dvd media?

**Comment by A1RF0RCE1** [comment]
I added this patch file : 

--- a/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp
+++ b/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp
@@ -167,7 +167,7 @@
   const AVCodec* codec = nullptr;
   void* i = 0;

-  if (!(hints.codecOptions & CODEC_FORCE_SOFTWARE))
+  if (!(hints.codecOptions & CODEC_FORCE_SOFTWARE)){
     while ((codec = av_codec_iterate(&i)))
     {
       if (!av_codec_is_decoder(codec))
@@ -179,7 +179,12 @@
       if (config)
         return codec;
     }
-
+  } else {
+      if ( hints.codec == 2){
+       CLog::Log(LOGINFO,"CDVDVideoCodecDRMPRIME::{} return nullptr for codec {}", __FUNCTION__, hints.codec);
+       return nullptr;
+    }
+  }


hints.codec == 2   is for dvd only ....

**Comment by popcornmix** [comment]
`hints.codec == 2` is `AV_CODEC_ID_MPEG2VIDEO`, so this will also disable hardware decode for (most) live TV, and some BluRays.

**Comment by A1RF0RCE1** [comment]
Thanks popcorn, I assumed mpeg2 was used for dvd only oops… is there a better way to accomplish this for the time being, until the menu overlay is fixed? The rpi4 is fast enough to decode DVD’s without hardware decode

**Comment by popcornmix** [comment]
https://github.com/xbmc/xbmc/blob/8cba5dc2145088e7c5667a75fd34685b0b2288e3/xbmc/cores/VideoPlayer/DVDCodecs/Video/DXVA.cpp#L1129 makes it slightly more specific, but still feels like a hack.
The HasMenu() in VideoPlayer.cpp feels like the right check, but whether that's valid at the time codec opens, I'm not sure.

Fixing rendering of the menu would obviously be the right solution. I can't think of a reason why it shouldn't work with drmprime, except we're triggering some bad behaviour through timing, or scaling.

I'd guess maybe menu is rendered but off-screen due to a scaling problem (e.g. similar to https://github.com/xbmc/xbmc/issues/20085) or perhaps the render region is incorrectly masked (e.g. viewport set with glScissor is making GUI not appear).

But I'm not too familiar with this code.

**Comment by richardshooteruk** [comment]
Hi, I've been asked to post a copy of [https://forum.libreelec.tv/thread/24517-when-using-a-different-screen-resolution-4k-to-to-a-dvd-dvd-navigation-and-subti/?postID=160443#post160443](url), I've included it below - in my case it is def. related to a difference between the Pi screen resolution and the DVD resolution:

Copy of original question/report:

Good morning,



first - thanks for providing this excellent media centre and all the work that went into and continues into it. The post is my experience of behaviour and may be intended though I suspect not, if not it may be useful for other people in the same edge case as me. Apologies if it is well known and understood - I could not find it.



The configuration in question involves:



Raspberry Pi 4 with 2Gb

LG C1 48" TV (3840 x 2160)

Filesystem images of DVD titles (complete DVD rip using dvdbackup)

Seen on 'Invasion of the Astro Monsters' Region 1 NTSC and "Pan's Labyrinth" and "City of the Lost Children Region" 2 PAL



Steps to reproduce:



1. Libreelec menus are set to a max 1080P resolution

2. Set the Pi screen resolution to 3840 x 2160

3. Open a DVD fileystem image (Video TS and VOBs etc) from the filesystem

4. Go to a DVD menu - no item is highlighted or the highlight is at the right edge of the screen off the menu items

5. Try and move round the menu with up / down / forward /back - nothing is high-lighted or it's at the right edge of the screen

6. Start a film and enable subtitles with 't'

7. Subtitles are missing or weirdly placed (and for the one DVD I could see them on a HUGE typeface!)



If the Pi screen res is set to 1920x1080 (standard HD) the DVD navigation is still a bit off (could be the DVDs, could be lovely new 10.0 build and port to Pi4) but the high lights are correctly positioned and subtitles are back in the lower middle of the screen.



I'd guess that the positioning is being done by screen size, the Pi isn't upscaling so they are off the edge of the screen (DVD images sends 576 wide or 1080p? highlights/sub-titles at ~1920/2 ?).



The workaround of setting screen size to 1080P works fine for me as I don't have any >HD content on libreelec at the moment, for people with 4k media it will presumably be a problem.



Happy to reproduce and provide logs, screenshots etc if it's considered a problem and investigation needed.



Thanks,



Richard

**Comment by popcornmix** [comment]
@A1RF0RCE1 can you make sure gui resolution and display resolution match and check if issue still occurs?

**Comment by A1RF0RCE1** [comment]
I ll check this first thing tomorrow!

**Comment by A1RF0RCE1** [comment]
These settings ain’t working in le 11 latest nightly. Screen is set to 1920/1080 menu 1080p also added the 1920/1080 in the config.txt no luck 😒

**Comment by popcornmix** [comment]
config.txt settings won't have an effect.

So these are the same?
settings/system/display/resolution
settings/system/display/Set gui resolution limit

Make sure both of those are 1080.
Also disable settings/player/adjust display refresh rate in case that is selecting a different hdmi resolution.

**Comment by A1RF0RCE1** [comment]
All settings are configured as you described, but it’s still not working… on a fresh start of libreelec you get a black screen when starting a dvd. you can select a title from the menu(in the blind) and the selected item will start and the movie shows up. Stopping the playblack and restart the movie the menu shows up but the highlighting does not ..

**Comment by popcornmix** [comment]
Okay, sounds like two issues, one with scaling and one with highlighting.

**Comment by A1RF0RCE1** [comment]
For now i added:
hints.codec == 2 && hints.height <= 576
Dvd resolution 720 * 480(ntsc) or 720 * 576 (Pal) this narrows it down a bit.

**Comment by ThexXTURBOXx** [comment]
I am also facing this issue in LE 10.0.0 which is pretty annoying.
Is there an estimated release date for the fix?

**Comment by git-developer** [comment]
I'm facing this issue on a Raspberry Pi 3B running OSMC 2021-11 (Kodi 19.3). What I found out:
* When DRM PRIME is disabled, both the DVD Menu and the cursor are shown correctly.
* When DRM PRIME is enabled and hardware acceleration is disabled, the DVD Menu is shown but the cursor is not shown.
* When DRM PRIME is enabled and hardware acceleration is enabled, the DVD Menu is not shown at all.

I compared the log of the Pi with the Log of a Windows machine, both opening the same DVD. The log does not show anything helpful; on the Windows machine, the menu is shown correctly and a few D3D messages are shown. On the Pi, the behavior is as described and no further messages are in the log.
